src/Queue/PhpRedis.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\redis\Queue\PhpRedis.
+ */
+
+namespace Drupal\redis\Queue;
+
+/**
+ * Redis queue implementation using PhpRedis extension backend.
+ *
+ * @ingroup queue
+ */
+class PhpRedis extends QueueBase {
+
+  /**
+   * The Redis connection.
+   *
+   * @var \Redis $client
+   */
+  protected $client;
+
+  /**
+   * Constructs a \Drupal\redis\Queue\PhpRedis object.
+   *
+   * @param string $name
+   *   The name of the queue.
+   * @param array $settings
+   *   Array of Redis-related settings for this queue.
+   * @param \Redis $client
+   *   The PhpRedis client.
+   */
+  function __construct($name, array $settings, \Redis $client) {
+    parent::__construct($name, $settings);
+    $this->client = $client;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function createItem($data) {
+    $record = new \stdClass();
+    $record->data = $data;
+    $record->qid = $this->incrementId();
+    // We cannot rely on REQUEST_TIME because many items might be created
+    // by a single request which takes longer than 1 second.
+    $record->timestamp = time();
+
+    $result = $this->client->multi()
```

Comment:
If you use either of MULTI, EXEC, DISCARD or WATCH, the code won't pass on shards, when using a proxy assisted shard either.

This makes it non compatible with proxy assisted sharding.